src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -429,6 +432,60 @@ GetShardCost(uint64 shardId, void *voidContext)
 }
 
 
+/*
+ * citus_partitioned_shard_total_size returns the total size of a partitioned table with
+ * its partitions, using the function pg_partition_tree.
+ */
+Datum
+citus_partitioned_shard_total_size(PG_FUNCTION_ARGS)
+{
+	char *shardName = text_to_cstring(PG_GETARG_TEXT_P(0));
+	uint64 size = 0;
+	const int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
+	Datum paramValues[1] = { CStringGetTextDatum(shardName) };
+
+	/* pg_partition_tree finds the partitions recursively and returns them */
+	char *queryText = "SELECT sum(pg_total_relation_size(relid))::bigint AS total_size "
+					  "FROM (SELECT relid from pg_partition_tree($1)) tables;";
```

Comment:
minor readability: `tables` -> `partition_tree`

And, it looks like this is the recommend way of calculating the size of the partition hierarchy -- by a PG committer: https://paquier.xyz/postgresql-2/postgres-12-partition-funcs/